arbor/simulation.cpp

```diff
@@ -186,6 +186,7 @@ simulation_state::simulation_state(
     task_system_(ctx.thread_pool),
     local_spikes_({thread_private_spike_store(ctx.thread_pool), thread_private_spike_store(ctx.thread_pool)})
 {
+    check_domain_decomposition(rec, ctx, decomp);
```

Comment:
In the C++ interface, `domain_decomposition` is a struct, and users can modify any part of that struct after construction. The check needs to happen here I think.

Reply:
We could stop that from happening. Is there ever a need for it?

Reply:
We could now that we have this function. It essentially does the same thing.

Reply:
I implemented the suggested change, but then I realized that I prefer to keep things as is for 2 reasons: 
- The struct is quite a jumble of things, and having to form it with a list initializer or a constructor will make the code more difficult to read.
- The constructor of `domain_decomposition` doesn't require the rec, or ctx. The partitioning algorithms do. So even if we make the changes to `domain_decomposition`, we still wouldn't be able to call `check_domain_decomposition` in the constructor of the class. 

What do you think?

Reply:
My reasoning was this:
- domain decompositions should be immutable after creation
- they should only be constructed via the algorithms (hard-ish to enforce, but doable)
- it's on the algorithms to check validity, they have also access to the relevant items (ctx et al)

Alternatively we could do this:
- simulations only work on checked_dom_dec
- checked_dom_dec can only be made by the check function

Reply:
For some context, the C++ `domain_decomposition` struct was designed for users to specify their own decomposition, or implement model-specific decomposition functions.
I don't think we want to restrict ourselves to the algorithms provided by Arbor, due to the difficulty of writing and maintaining algorithms that are applicable to all problems.

A stand-alone validator for domain decompositions.

In Python land the story is different: it isn't possible to construct a domain decomposition in Python because it requires writing a gid mapping function, which would mean supporting Python callbacks, which have serious consequences for parallelisation.

So, if a domain decomposition is created by one of the algorithms provided via the Python interface, they should validate the decomposition.

But I don't think we need to make this a hard-coded requirement on the C++ side.

Reply:
For some context, the C++ domain_decomposition struct was designed for users to specify their own decomposition, or implement model-specific decomposition functions.
I don't think we want to restrict ourselves to the algorithms provided by Arbor, due to the difficulty of writing and maintaining algorithms that are applicable to all problems.

A stand-alone validator for domain decompositions.

In Python land the story is different: it isn't possible to construct a domain decomposition in Python because it requires writing a gid mapping function, which would mean supporting Python callbacks, which have serious consequences for parallelisation.

So, if a domain decomposition is created by one of the algorithms provided via the Python interface, they should validate the decomposition.

But I don't think we need to make this a hard-coded requirement on the C++ side.

Reply:
Hey @bcumming thanks for the context.

> For some context, the C++ domain_decomposition struct was designed for users to specify their own decomposition, or implement model-specific decomposition functions.
I don't think we want to restrict ourselves to the algorithms provided by Arbor, due to the difficulty of writing and maintaining algorithms that are applicable to all problems.

That's great, so this data (`domain_decomposition`) is a piece of user-supplied input and we have seen (ref one-sided gap-junctions) that it is very well possible to produce strange behaviour if used incorrectly. I think it's good practice to validate this such that
1. it is checked early in the flow in order to get feedback to the user sooner
2. it cannot be changed afterwards without re-validation
3. Preferably enforce 2. using the type system

The current state is to check upon construction of the simulation, which is fine, but it delays user feedback quite a bit.
That was the core of my suggestion.

> But I don't think we need to make this a hard-coded requirement on the C++ side.

Even in C++ code decompositions made by us should be valid. Else the validation scheme is broken and needs to be fixed.

Reply:
I think the key point in Ben's comment is: 
> the C++ domain_decomposition struct was designed for users to specify their own decomposition, or implement model-specific decomposition functions.
I don't think we want to restrict ourselves to the algorithms provided by Arbor, due to the difficulty of writing and maintaining algorithms that are applicable to all problems.

If we agree that `domain_decomposition` should remain as it is in order to allow users to write their own partitioning algorithms, then we have given the users the tool to check their structs as early on as they want using the public `check_domain_decomposition` function. If the users neglect to do that, it will be checked again before the object is used in the constructor of `simulation`. 

However, we could argue that the users can use the newly added `partition_by_group` to create their own custom `domain_decomposition` objects and therefore the `domain_decomposition` struct doesn't need to remain as it is. The key bit of information that the user would be interested in (the different cell groupings on each rank) is passed as an argument to `partition_by_group` which then fills out the rest of the struct members. So, we could make the `domain_decomposition` struct more opaque to the users. We could make `partition_by_group` the only constructor of `domain_decomposition` (so `partition_load_balance` would need to change a bit), and we can perform the correctness check in the constructor.

Reply:
👍 That's essentially what I had in mind, but was unsure whether `partition_by_group` would be the only viable (smart) constructor.